### PR TITLE
drivers: display: Add ili9163

### PR DESCRIPTION
**Pull request by pavlohamov**
Implement driver for Ilitek ili9163

### CODE REVIEW

**Review by jfischer-no — CHANGES_REQUESTED** [review]
Please follow [Coding Style](https://docs.zephyrproject.org/latest/contribute/index.html#coding-style). Please take a look at new [GPIO API](https://github.com/zephyrproject-rtos/zephyr/tree/topic-gpio). This driver needs at least a generic shield like [st7789v_generic](https://github.com/zephyrproject-rtos/zephyr/tree/topic-gpio/boards/shields/st7789v_generic).

**Review by jfischer-no — CHANGES_REQUESTED** [review]
Why do we need a driver for ILI9163 controller and can not enhance/use ILI9340?

Before that, ILI9340 driver needs some rework to obtain all the panel parameters from DT, like it was done for ST7789V.

File: drivers/display/display_ili9163.c [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+
+#include <drivers/gpio.h>
+#include <sys/byteorder.h>
+#include <drivers/spi.h>
+#include <string.h>
+
+#define ILI9163_X_MAX 162U
+#define ILI9163_Y_MAX 132U
+
+#define ILI9163_CMD_DATA_PIN_COMMAND 0
```
[the remark this answers is not in the document]
Comment (pavlohamov) [reply drivers/display/display_ili9163.c]: Make sense. removed

File: drivers/display/display_ili9163.c [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+	(DT_INST_0_ILITEK_ILI9163_X_OFFSET > (ILI9163_X_MAX - RES_HOR))
+#error "Please check x-offset of \"ilitek,ili9163\" device"
+#endif
+
+#if defined(DT_INST_0_ILITEK_ILI9163_Y_OFFSET) && \
+	(DT_INST_0_ILITEK_ILI9163_Y_OFFSET > (ILI9163_Y_MAX - RES_VER))
+#error "Please check y-offset of \"ilitek,ili9163\" device"
+#endif
+
+#ifdef CONFIG_LVGL
```
[the remark this answers is not in the document]
Comment (pavlohamov) [reply drivers/display/display_ili9163.c]: Done

File: drivers/display/display_ili9163.c [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+	struct spi_cs_control cs_ctrl;
+#endif
+	enum display_orientation orientation;
+	u8_t height;
+	u8_t width;
+	u8_t x_offset;
+	u8_t y_offset;
+};
+
+static int ili9163_blanking_off(const struct device *dev);
```
[the remark this answers is not in the document]
Comment (pavlohamov) [reply drivers/display/display_ili9163.c]: Oups, i didn't event think about moving. thank you for suggestion. Done

File: drivers/display/display_ili9163.c [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+	cmd = DT_INST_0_ILITEK_ILI9163_MADC;
+#else
+	cmd = ILI9163_DATA_MEM_ACCESS_CTRL_BGR;
+#endif
+	ili9163_transmit(data, ILI9163_CMD_MEM_ACCESS_CTRL, &cmd, sizeof(cmd));
+
+	ili9163_transmit(data, ILI9163_CMD_MOD_NORMAL, NULL, 0);
+}
+
+static int ili9163_init(struct device *dev)
```
[the remark this answers is not in the document]
Comment (pavlohamov) [reply drivers/display/display_ili9163.c]: Done

File: drivers/display/display_ili9163.c [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+static int ili9163_set_brightness(const struct device *dev,
+				  const u8_t brightness)
+{
+	LOG_WRN("Set brightness not supported");
+	return -ENOTSUP;
+}
+
+static int ili9163_set_contrast(const struct device *dev, const u8_t contrast)
+{
+	LOG_ERR("Set contrast not supported");
```
[the remark this answers is not in the document]
Comment (pavlohamov) [reply drivers/display/display_ili9163.c]: My bad. Thanks, fixed

File: drivers/display/display_ili9163.h [hunk 6]
```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) 2020 Jabil Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_DRIVERS_DISPLAY_DISPLAY_ILI9163_H_
+#define ZEPHYR_DRIVERS_DISPLAY_DISPLAY_ILI9163_H_
+
+#define ILI9163_CMD_SOFTWARE_RESET 0x01
```
[the remark this answers is not in the document]
Comment (pavlohamov) [reply drivers/display/display_ili9163.h]: Done

File: drivers/display/display_ili9163.c [hunk 7]
```diff
@@ -0,0 +1,414 @@
...
+	return -ENOTSUP;
+}
+
+static int ili9163_set_pixel_format(const struct device *dev,
+				    const enum display_pixel_format
+				    pixel_format)
+{
+#ifdef CONFIG_ILI9163_RGB565
```
[the remark this answers is not in the document]
Comment (pavlohamov) [reply drivers/display/display_ili9163.c]: Done

File: drivers/display/display_ili9163.c [hunk 8]
```diff
@@ -0,0 +1,414 @@
...
+
+static void ili9163_get_capabilities(const struct device *dev,
+				     struct display_capabilities *capabilities)
+{
+	struct ili9163_data *data = (struct ili9163_data *)dev->driver_data;
+
+	memset(capabilities, 0, sizeof(struct display_capabilities));
+	capabilities->x_resolution = RES_HOR;
+	capabilities->y_resolution = RES_VER;
+#ifdef CONFIG_ILI9163_RGB565
```
[the remark this answers is not in the document]
Comment (pavlohamov) [reply drivers/display/display_ili9163.c]: Done

File: drivers/display/display_ili9163.c [hunk 9]
```diff
@@ -0,0 +1,414 @@
...
+static inline void ili9163_exit_sleep(struct ili9163_data *data)
+{
+	ili9163_transmit(data, ILI9163_CMD_EXIT_SLEEP, NULL, 0);
+	k_sleep(K_MSEC(5));
+}
+
+static inline void ili9163_lcd_init(struct ili9163_data *data)
+{
+	u8_t cmd = 0;
+#ifdef CONFIG_ILI9163_RGB565
```
[the remark this answers is not in the document]
Comment (pavlohamov) [reply drivers/display/display_ili9163.c]: > 
> 
> Why do we need a driver for ILI9163 controller and can not enhance/use ILI9340?
> 
> Before that, ILI9340 driver needs some rework to obtain all the panel parameters from DT, like it was done for ST7789V.

Good idea. As for now i've fixed your comments. I will see what i can do to combine them

**Comment by pavlohamov** [comment]
> 
> 
> Why do we need a driver for ILI9163 controller and can not enhance/use ILI9340?
> 
> Before that, ILI9340 driver needs some rework to obtain all the panel parameters from DT, like it was done for ST7789V.

Good idea. As for now i've fixed your comments. I will see what i can do to combine them

**Comment by pavlohamov** [comment]
@jfischer-phytec-iot Could you be so kind to take a look on https://github.com/zephyrproject-rtos/zephyr/pull/22674 ?

**Pull request closed without merging**